### PR TITLE
Prohibit non-ASCII characters in URLs, nudge toward punycode

### PR DESCRIPTION
**Pull request by colindean**
Inspired by curl's blog post, [Detecting malicious Unicode][1], this likely captures most if not all cases and nudges the user toward supplying IDNs with punycode.

A possible improvement would be telling the user exactly what punycode domain to use instead, but that may require another library as I can't quickly find something built into the Ruby stdlib that handles punycode encoding.

This also needs tests, to follow.

[1]: https://daniel.haxx.se/blog/2025/05/16/detecting-malicious-unicode/

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/rubocops/shared/url_helper.rb [hunk 1]
```diff
@@ -35,6 +35,12 @@ def audit_urls(urls, regex)
       def audit_url(type, urls, mirrors, livecheck_url: false)
         @type = type
 
+        # URLs must be ASCII; IDNs must be punycode
+        ascii_pattern = %r{[^\p{ASCII}]+}
```
Comment (colindean) [inline Library/Homebrew/rubocops/shared/url_helper.rb:39]: **chore:** 
```suggestion
        ascii_pattern = /[^\p{ASCII}]+/
```

File: Library/Homebrew/rubocops/shared/url_helper.rb [hunk 2]
```diff
@@ -35,6 +35,12 @@ def audit_urls(urls, regex)
       def audit_url(type, urls, mirrors, livecheck_url: false)
         @type = type
 
+        # URLs must be ASCII; IDNs must be punycode
+        ascii_pattern = /[^\p{ASCII}]+/
+        audit_urls(URLs, ascii_pattern) do |_, url|
```
Comment (colindean) [inline Library/Homebrew/rubocops/shared/url_helper.rb:40]: **typo:** Stupid autocorrect in GH editor…
```suggestion
        audit_urls(urls, ascii_pattern) do |_, url|
```

**Comment by colindean** [comment]
Failure is unrelated to my addition.

File: Library/Homebrew/rubocops/shared/url_helper.rb [hunk 3]
```diff
@@ -35,6 +35,12 @@ def audit_urls(urls, regex)
       def audit_url(type, urls, mirrors, livecheck_url: false)
         @type = type
 
+        # URLs must be ASCII; IDNs must be punycode
+        ascii_pattern = /[^\p{ASCII}]+/
```
Comment (colindean) [inline Library/Homebrew/rubocops/shared/url_helper.rb:39]: **issue:** Potentially, non-ASCII may be acceptable in path or query components of the URL. This would fail those.

**Comment by colindean** [comment]
Test failure will be fixed by https://github.com/Homebrew/brew/pull/19982.

File: Library/Homebrew/rubocops/shared/url_helper.rb [hunk 4]
```diff
@@ -35,6 +35,12 @@ def audit_urls(urls, regex)
...
+        audit_urls(urls, ascii_pattern) do |_, url|
+          problem "Please use the punycode version of #{url}."
```
Comment (stefanb) [inline Library/Homebrew/rubocops/shared/url_helper.rb:41]:
```suggestion
          problem "Please use the ASCII (punycode, URL encodeded) version of #{url}."
```

Comment (stefanb) [reply Library/Homebrew/rubocops/shared/url_helper.rb:41]: If we want to be really helpful we would validate the hostname and path separately, and write the correct error message and correctly encoded expected full URL.

File: Library/Homebrew/rubocops/shared/url_helper.rb [hunk 5]
```diff
@@ -35,6 +35,12 @@ def audit_urls(urls, regex)
       def audit_url(type, urls, mirrors, livecheck_url: false)
         @type = type
 
+        # URLs must be ASCII; IDNs must be punycode
+        ascii_pattern = /[^\p{ASCII}]+/
+        audit_urls(urls, ascii_pattern) do |_, url|
+          problem "Please use the ASCII (punycode, URL encodeded) version of #{url}."
```
Comment (colindean) [inline Library/Homebrew/rubocops/shared/url_helper.rb:41]: **typo:** 
```suggestion
          problem "Please use the ASCII (Punycode, URL encoded) version of #{url}."
```

File: Library/Homebrew/test/rubocops/urls_spec.rb [hunk 6]
```diff
@@ -177,6 +177,14 @@
       "url" => "svn+http://brew.sh/foo/bar",
       "msg" => "Use of the svn+http:// scheme is deprecated, pass `:using => :svn` instead",
       "col" => 2,
+    }, {
+      "url" => "https://🫠.sh/foo/bar",
+      "msg" => "Please use the punycode version of https://🫠.sh/foo/bar.",
+      "col" => 2,
+    }, {
+      "url" => "https://ßreｗ.sh/foo/bar",
+      "msg" => "Please use the punycode version of https://ßreｗ.sh/foo/bar.",
```
Comment (colindean) [inline Library/Homebrew/test/rubocops/urls_spec.rb:186]: **chore:** 
```suggestion
      "msg" => "Please use the ASCII (Punycode, URL encoded) version of https://🫠.sh/foo/bar.",
      "col" => 2,
    }, {
      "url" => "https://ßreｗ.sh/foo/bar",
      "msg" => "Please use the ASCII (Punycode, URL encoded) version of https://ßreｗ.sh/foo/bar.",
```

File: Library/Homebrew/test/rubocops/urls_spec.rb [hunk 7]
```diff
@@ -177,6 +177,14 @@
       "url" => "svn+http://brew.sh/foo/bar",
       "msg" => "Use of the svn+http:// scheme is deprecated, pass `:using => :svn` instead",
       "col" => 2,
+    }, {
+      "url" => "https://🫠.sh/foo/bar",
+      "msg" => "Please use the ASCII (Punycode, URL encoded) version of https://🫠.sh/foo/bar.",
+      "col" => 2,
+    }, {
+      "url" => "https://ßreｗ.sh/foo/bar",
```
Comment (stefanb) [inline Library/Homebrew/test/rubocops/urls_spec.rb:185]: Few more test cases for unicode path & query:
* `https://example.com/ßreｗ`
* `https://example.org/?project=ßreｗ&foo=bar`

File: Library/Homebrew/rubocops/shared/url_helper.rb [hunk 8]
```diff
@@ -35,6 +35,12 @@ def audit_urls(urls, regex)
       def audit_url(type, urls, mirrors, livecheck_url: false)
         @type = type
 
+        # URLs must be ASCII; IDNs must be punycode
+        ascii_pattern = /[^\p{ASCII}]+/
+        audit_urls(urls, ascii_pattern) do |_, url|
+          problem "Please use the ASCII (Punycode, URL encoded) version of #{url}."
```
Comment (stefanb) [inline Library/Homebrew/rubocops/shared/url_helper.rb:41]:
```suggestion
          problem "Please use the ASCII (Punycode host, URL encoded path) version of #{url}."
```

**Comment by gromgit** [comment]
If a correctly-encoded URL can be derived, it also sounds like a candidate for auto-correct (i.e. `brew audit --fix`).

File: Library/Homebrew/rubocops/shared/url_helper.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/rubocops/shared/url_helper.rb:39]: I'd rather we be excessively strict for now, provided homebrew/core and homebrew/cask pass, and loosen it later.

File: Library/Homebrew/test/rubocops/urls_spec.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/rubocops/urls_spec.rb:185]: Let's punt on different errors for now.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @colindean! Looks good. Thanks @stefanb for the great review too. Let's get this landed ASAP and iterate.

File: Library/Homebrew/test/rubocops/urls_spec.rb [hunk 9]
```diff
@@ -177,6 +177,14 @@
       "url" => "svn+http://brew.sh/foo/bar",
       "msg" => "Use of the svn+http:// scheme is deprecated, pass `:using => :svn` instead",
       "col" => 2,
+    }, {
+      "url" => "https://🫠.sh/foo/bar",
+      "msg" => "Please use the ASCII (Punycode, URL encoded) version of https://🫠.sh/foo/bar.",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/rubocops/urls_spec.rb:182]:
```suggestion
      "msg" => "Please use the ASCII (Punycode encoded host, URL-encoded path and query) version of https://🫠.sh/foo/bar.",
```

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/test/rubocops/urls_spec.rb [hunk 10]
```diff
@@ -177,6 +177,14 @@
       "url" => "svn+http://brew.sh/foo/bar",
       "msg" => "Use of the svn+http:// scheme is deprecated, pass `:using => :svn` instead",
       "col" => 2,
+    }, {
+      "url" => "https://🫠.sh/foo/bar",
+      "msg" => "Please use the ASCII (Punycode encoded host, URL-encoded path and query) version of https://🫠.sh/foo/bar.",
+      "col" => 2,
+    }, {
+      "url" => "https://ßreｗ.sh/foo/bar",
+      "msg" => "Please use the ASCII (Punycode, URL encoded) version of https://ßreｗ.sh/foo/bar.",
```
Comment (stefanb) [inline Library/Homebrew/test/rubocops/urls_spec.rb:186]: The old message was forgotten in this test:
```suggestion
      "msg" => "Please use the ASCII (Punycode encoded host, URL-encoded path and query) version of https://ßreｗ.sh/foo/bar.",
```

**Comment by MikeMcQuaid** [comment]
Thanks @colindean great work!
